packages/z2s/src/test/chinook/chinook.pg-test.ts

```diff
@@ -185,7 +219,221 @@ async function checkZqlAndSql(
+    return new Map(
+      wrapIterable(queryRows.entries()).map(([table, rows]) => [
+        table,
+        [...rows.values()],
+      ]),
+    );
+  }
+
+  const totalNumRows = [...queryRows.values()].reduce(
+    (acc, rows) => acc + rows.size,
+    0,
+  );
+
+  const interval = Math.floor(totalNumRows / 50);
+  const removedRows = await checkRemove(
+    interval,
+    copyRows(),
+    pg,
+    zqliteQuery,
+    memoryQuery,
+  );
+  await checkAddBack(removedRows, pg, zqliteQuery, memoryQuery);
+  checkEditToRandom();
+  checkEditToMatch();
+}
+
+// Removes all rows that are in the result set
+// one at a time till there are no rows left.
+// Randomly selects which table to remove a row from on each iteration.
+async function checkRemove(
+  removalInterval: number,
+  queryRows: Map<string, Row[]>,
+  sql: PostgresDB,
+  zqliteQuery: Query<Schema, keyof Schema['tables']>,
+  memoryQuery: Query<Schema, keyof Schema['tables']>,
+): Promise<[table: string, row: Row][]> {
+  const tables = [...queryRows.keys()];
+
+  const zqliteMaterialized = zqliteQuery.materialize();
+  const zqlMaterialized = memoryQuery.materialize();
+  const sqlQuery = formatPg(compile(ast(zqliteQuery), format(zqliteQuery)));
+
+  let numOps = 0;
+  const removedRows: [string, Row][] = [];
+  const removals = new Set<string>();
+  while (tables.length > 0) {
+    ++numOps;
+    const tableIndex = Math.floor(Math.random() * tables.length);
+    const table = tables[tableIndex];
+    const rows = must(queryRows.get(table));
+    const rowIndex = Math.floor(Math.random() * rows.length);
+    const row = must(rows[rowIndex]);
+    rows.splice(rowIndex, 1);
+
+    if (rows.length === 0) {
+      tables.splice(tableIndex, 1);
+    }
+
+    // doing this for all rows of a large table
+    // is too slow we only do it every `removalInterval`
+    if (numOps % removalInterval === 0) {
+      removedRows.push([table, row]);
+      const {primaryKey} = schema.tables[table as keyof Schema['tables']];
+      await sql`DELETE FROM ${sql(table)} WHERE ${primaryKey
+        .map(col => sql`${sql(col)} = ${row[col] ?? null}`)
+        .reduce((l, r) => sql`${l} AND ${r}`)}`;
+
+      try {
+        must(zqliteQueryDelegate.getSource(table)).push({
+          type: 'remove',
+          row,
+        });
+        must(memoryQueryDelegate.getSource(table)).push({
+          type: 'remove',
+          row,
+        });
+        removals.add(`${table}-${JSON.stringify(row)}`);
+      } catch (e) {
+        throw new Error(
+          `Failed removal of ${JSON.stringify(
+            row,
+          )} for ${table} contained ${removals.has(
+            `${table}-${JSON.stringify(row)}`,
+          )}`,
+        );
+      }
+
+      const pgResult = await sql.unsafe(
+        sqlQuery.text,
+        sqlQuery.values as JSONValue[],
+      );
+      // TODO: relationships return `undefined` from ZQL and `null` from PG
+      expect(zqliteMaterialized.data).toEqualNullish(pgResult);
```

Comment:
`toEqualNullish` is because ZQL is returning `undefined` for empty `one` relationships and `z2s` returns null.